osquery/core/watcher.cpp

```diff
@@ -345,21 +345,29 @@ bool WatcherRunner::watch(const PlatformProcess& child) const {
 }
 
 void WatcherRunner::stopChild(const PlatformProcess& child) const {
-  child.killGracefully();
+  const auto childPid = child.pid();
+  if (childPid < 0) {
+    const auto message = std::string("Watcher cannot stop worker process (") + std::to_string(childPid) + ").";
+    LOG(INFO) << message;
+    return;
+  }
 
-  // Clean up the defunct (zombie) process.
-  if (!child.cleanup()) {
-    auto child_pid = child.pid();
+  const auto killedGracefully = child.killGracefully();
+  if (killedGracefully) {
```

Comment:
This fix is not correct for *nix systems.

bool PlatformProcess::killGracefully() const {
  if (!isValid()) {
    return false;
  }

  int status = ::kill(nativeHandle(), SIGINT);
  return (status == 0);
}

Sending a signal to a process on a *nix is async affair. So even if kill return successfully does not mean that process is terminated it only means that signal is delivered to that process so when the cpu scheduler picks up that process to run again it check if a signal pending for that process and based upon the signal disposition things happen( i.e. calling signal handler  or termination or ignoring signal)

Reply:
Interesting. Thanks for the feedback! If there's so much difference between *nix systems and Windows does it make sense to use the same function to shutdown/cleanup? Maybe there should be 2 different implementations?

Reply:
Cleanup(child.cleanup()) is actually waiting(few seconds here in a loop) to confirm the termination of the child (worker) and reading its exit status  failure to do so will leave died worker process leave in zombie state unless its parent (watcher) dies as well.